### PR TITLE
feat(openid-connect): Add `symmetric_key` option to support verifying tokens with symmetric algorithms

### PR DESCRIPTION
**Pull request by liweitianux**
### Description

The underlying `lua-resty-openidc` module already supports the
`symmetric_key` option to specify the HMAC key for verifying HS???
tokens.  However, note that `lua-resty-openidc` just passes the
`symmetric_key` value as-is to HMAC.  So we choose to accept a
base64url-encoded secret, which is easier to obtain from OP and
configure, and then decode it before passing it to `lua-resty-openidc`.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by liweitianux** [comment]
**NOTE**: the `lua-resty-openidc` is having a bug in verifying HS??? tokens, and I've submitted a PR to fix it: https://github.com/zmartzone/lua-resty-openidc/pull/447

File: apisix/plugins/openid-connect.lua [hunk 1]
```diff
@@ -71,6 +72,10 @@ local schema = {
             type = "string",
             description = "the URI will be redirect when request logout_path",
         },
+        symmetric_key = {
+            type = "string",
+            description = "base64url-encoded secret for verifying HS??? tokens",
```
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:77]: What's `HS???` token?

**Review by spacewander — COMMENTED** [review]
Could you add some tests for this feature? How can we test it?

File: apisix/plugins/openid-connect.lua (hunk 1 above)
Comment (liweitianux) [reply apisix/plugins/openid-connect.lua:77]: like HS256, HS512

**Comment by liweitianux** [comment]
> Could you add some tests for this feature? How can we test it?

Hmm, I'll have to first check out the existing tests (e.g., for RS256) and figure out one for the HS256/HS512 tokens.

File: apisix/plugins/openid-connect.lua (hunk 1 above)
Comment (tzssangglass) [reply apisix/plugins/openid-connect.lua:77]: we can just use `HS256, HS512`, not `HS???`,  it looks unclear

Comment (liweitianux) [reply apisix/plugins/openid-connect.lua:77]: ok. sure :)

**Comment by liweitianux** [comment]
> **NOTE**: the `lua-resty-openidc` is having a bug in verifying HS??? tokens, and I've submitted a PR to fix it: [zmartzone/lua-resty-openidc#447](https://github.com/zmartzone/lua-resty-openidc/pull/447)

Well. `lua-resty-openidc` is actually good. It's the Keycloak (I was testing with) having the issue with HS256 (and other HS algorithms) tokens.

By the way, I'll have to consider how to best pass the `symmetric_key` option (e.g., whether to encode or not).

**Comment by spacewander** [comment]
> It's the Keycloak (I was testing with) having the issue with HS256 (and other HS algorithms) tokens.

Is there any link for this issue in Keycloak? Would you explain it in depth?

**Comment by spacewander** [comment]
> > It's the Keycloak (I was testing with) having the issue with HS256 (and other HS algorithms) tokens.
> 
> Is there any link for this issue in Keycloak? Would you explain it in depth?

Solved it: https://github.com/keycloak/keycloak/issues/13823.

File: apisix/plugins/openid-connect.lua (hunk 1 above)
Comment (kayx23) [reply apisix/plugins/openid-connect.lua]: I updated

**Comment by kayx23** [comment]
Re-opened as this is a meaningful contribution.

Helped resolve conflicts and update the explanation for the `symmetric_key` to be a bit more readable, because the PR author mentioned in a different PR they've changed jobs and might not be able to continue the effort here, especially with adding test cases.

**Comment by Revolyssup** [comment]
@liweitianux Can you no longer work on this PR?

**Comment by liweitianux** [comment]
> @liweitianux Looks like you're busy. We will assign this issue to someone else.

Hi @Revolyssup, sorry that I haven't been using APISIX for some time, and I currently don't have the time for this PR.  Please help reassign it. Thanks.

**Pull request closed without merging**